### PR TITLE
[Core] Remove promotion and tax adjustments before recalculation

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Related tickets |
| License         | MIT

### CODE REVIEW

**Comment by peteward** [comment]
looks good to me 👍

File: src/Sylius/Component/Core/Remover/AdjustmentsRemover.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class AdjustmentsRemover implements AdjustmentsRemoverInterface
+{
+    /**
+     * @var array
+     */
+    protected $adjustmentsToRemove = [
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Remover/AdjustmentsRemover.php]: Why protected?

**Comment by mmckelviereiss** [comment]
There needs to be a change to drop the removal of adjustments in the various processors. As one example, we can drop the `OrderTaxesProcessor::clearTaxes` private method.

File: src/Sylius/Component/Core/Remover/AdjustmentsRemoverInterface.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface AdjustmentsRemoverInterface
+{
+    /**
+     * @param OrderInterface $order
+     */
+    public function remove(OrderInterface $order);
```
Comment (pamil) [inline src/Sylius/Component/Core/Remover/AdjustmentsRemoverInterface.php]: `remove` -> `removeFrom` / `removeFromOrder`?

File: src/Sylius/Component/Core/Remover/AdjustmentsRemover.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function remove(OrderInterface $order)
+    {
+        $units = $order->getItemUnits();
+        foreach ($this->adjustmentsToRemove as $type) {
+            $order->removeAdjustments($type);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Remover/AdjustmentsRemover.php]: How about introducing `removeAdjustmentsRecursively($type)` that will clear all adjustments of given type on all levels? Right now we remove only adjustments from `Order` and `OrderItemUnits`, but we could have adjustments applied on `OrderItem` level also.

File: src/Sylius/Component/Core/spec/Remover/AdjustmentsRemoverSpec.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    function it_removes_adjustments_from_order_recursively(
+        OrderInterface $order,
+        OrderItemUnitInterface $firstUnit,
+        OrderItemUnitInterface $secondUnit
+    ) {
+        $order->removeAdjustments(AdjustmentInterface::ORDER_PROMOTION_ADJUSTMENT)->shouldBeCalled();
+        $order->removeAdjustments(AdjustmentInterface::ORDER_ITEM_PROMOTION_ADJUSTMENT)->shouldBeCalled();
+        $order->removeAdjustments(AdjustmentInterface::TAX_ADJUSTMENT)->shouldBeCalled();
+
+        $order->getItemUnits()->willReturn(new \ArrayIterator([$firstUnit->getWrappedObject(), $secondUnit->getWrappedObject()]));
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Remover/AdjustmentsRemoverSpec.php]: `ArrayIterator` -> `ArrayCollection` :)

**Comment by Zales0123** [comment]
@mmckelviereiss Surely, we should **rethink** order/taxes/promotion recalculation logic, however, I think such a methods can be left for now - services like ``OrderTaxesProcessor`` could be somehow used independently, not by ``OrderRecalculator``. Fortunately it's not a big performance issue (only looping on empty collection).

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! 👍

**Pull request merged**